euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/ReduceStateByKeyTranslator.java

```diff
@@ -44,114 +47,124 @@
   public DataStream<?> translate(FlinkOperator<ReduceStateByKey> operator,
                                  StreamingExecutorContext context)
   {
-    DataStream<?> input =
+    DataStream input =
             Iterables.getOnlyElement(context.getInputStreams(operator));
 
     ReduceStateByKey origOperator = operator.getOriginalOperator();
 
     StateFactory<?, State> stateFactory = origOperator.getStateFactory();
+    CombinableReduceFunction stateCombiner = origOperator.getStateCombiner();
+
+    Windowing windowing = origOperator.getWindowing();
+    if (windowing == null) {
+      // use attached windowing when no windowing explicitly defined
+      windowing = new AttachedWindowing<>();
+    }
 
-    final Windowing windowing = origOperator.getWindowing();
     final UnaryFunction keyExtractor = origOperator.getKeyExtractor();
     final UnaryFunction valueExtractor = origOperator.getValueExtractor();
     final UnaryFunction eventTimeAssigner = origOperator.getEventTimeAssigner();
 
-    FlinkStreamingStateStorageProvider storageProvider
-        = new FlinkStreamingStateStorageProvider();
-
-    DataStream<StreamingWindowedElement<?, Pair>> folded;
-    // apply windowing first
-    if (windowing == null) {
-      WindowedStream windowed =
-          context.attachedWindowStream((DataStream) input, keyExtractor, valueExtractor);
-      // equivalent operation to "left fold"
-      folded = windowed.apply(new RSBKWindowFunction(storageProvider, stateFactory))
-          .name(operator.getName())
-          .setParallelism(operator.getParallelism());
-    } else {
-      WindowedStream<MultiWindowedElement<?, Pair>, Object, FlinkWindow>
-          windowed = context.flinkWindow(
-              (DataStream) input, keyExtractor, valueExtractor, windowing, eventTimeAssigner);
-      // equivalent operation to "left fold"
-      folded = windowed.apply(new RSBKWindowFunction(storageProvider, stateFactory))
-          .name(operator.getName())
-          .setParallelism(operator.getParallelism());
+    if (eventTimeAssigner != null) {
+      input = input.assignTimestampsAndWatermarks(
+              new EventTimeAssigner(context.getAllowedLateness(), eventTimeAssigner));
     }
 
+    // assign windows
+    DataStream<KeyedMultiWindowedElement> windowed = input.map(new WindowAssigner(
+            windowing, keyExtractor, valueExtractor, eventTimeAssigner))
+            .setParallelism(operator.getParallelism());
+
+    DataStream<WindowedElement<?, Pair>> reduced = (DataStream) windowed.keyBy(new KeyExtractor())
+            .transform(operator.getName(), TypeInformation.of(WindowedElement.class), new WindowOperator<>(
+                    windowing, stateFactory, stateCombiner, context.isLocalMode()))
+            .setParallelism(operator.getParallelism());
```

Comment:
i would like to see support for running this "as is", as well as with the value extraction and window assignment functionality executing only after the shuffle (maybe some global parameter to the translation layer). at least for our benchmarking this will be necessary. what do you think it would take to support both?